scalding-core/src/main/scala/com/twitter/scalding/macros/impl/FieldsProviderImpl.scala

```diff
@@ -80,53 +80,81 @@ object FieldsProviderImpl {
-        This will mean the macro is operating on a non-resolved type.""")
-
-    def matchField(fieldType: Type, outerName: Option[String], fieldName: String, isOption: Boolean): List[(Tree, String)] = {
-      val returningType = if (isOption) q"""classOf[java.lang.Object]""" else q"""classOf[$fieldType]"""
-      val simpleRet = outerName match {
-        case Some(outer) => List((returningType, s"$outer$fieldName"))
-        case None => List((returningType, s"$fieldName"))
-      }
-      fieldType match {
-        case tpe if tpe =:= typeOf[String] => simpleRet
-        case tpe if tpe =:= typeOf[Boolean] => simpleRet
-        case tpe if tpe =:= typeOf[Short] => simpleRet
-        case tpe if tpe =:= typeOf[Int] => simpleRet
-        case tpe if tpe =:= typeOf[Long] => simpleRet
-        case tpe if tpe =:= typeOf[Float] => simpleRet
-        case tpe if tpe =:= typeOf[Double] => simpleRet
-        case tpe if tpe.erasure =:= typeOf[Option[Any]] && isOption == true => c.abort(c.enclosingPosition, s"Case class ${T} has nested options, not supported currently.")
-        case tpe if tpe.erasure =:= typeOf[Option[Any]] =>
-          val innerType = tpe.asInstanceOf[TypeRefApi].args.head
-          matchField(innerType, outerName, fieldName, true)
-        case tpe if IsCaseClassImpl.isCaseClassType(c)(tpe) =>
-          val prefix = outerName.map(pre => s"$pre$fieldName.")
-          expandMethod(tpe, prefix, isOption)
-        case tpe if allowUnknownTypes => simpleRet
-        case _ => c.abort(c.enclosingPosition, s"Case class ${T} is not pure primitives or nested case classes")
+    def maybeHandlePrimitive(outerTpe: Type): Option[List[(Tree, String)]] = {
+
+      def innerLoop(tpe: Type, inOption: Boolean): Option[List[(Tree, String)]] = {
+        val returningType = if (inOption) q"""classOf[java.lang.Object]""" else q"""classOf[$outerTpe]"""
+        val simpleRet = Some(List((returningType, s"element")))
+
+        tpe match {
+          case tpe if tpe =:= typeOf[String] => simpleRet
+          case tpe if tpe =:= typeOf[Boolean] => simpleRet
+          case tpe if tpe =:= typeOf[Short] => simpleRet
+          case tpe if tpe =:= typeOf[Int] => simpleRet
+          case tpe if tpe =:= typeOf[Long] => simpleRet
+          case tpe if tpe =:= typeOf[Float] => simpleRet
+          case tpe if tpe =:= typeOf[Double] => simpleRet
+          case tpe if tpe.erasure =:= typeOf[Option[Any]] && inOption == true => c.abort(c.enclosingPosition, s"Case class ${T} has nested options, not supported currently.")
+          case tpe if tpe.erasure =:= typeOf[Option[Any]] =>
+            val innerType = tpe.asInstanceOf[TypeRefApi].args.head
+            innerLoop(innerType, true)
+
+          case tpe => None
+        }
       }
+      innerLoop(outerTpe, false)
     }
+    val expanded: List[(Tree, String)] = maybeHandlePrimitive(T.tpe).getOrElse {
```

Comment:
I think we should have `maybeHandlePrimitive` return `Option[Tree]` then here we can do:

```scala
val expanded: List[(Tree, Comparable[_])]  = maybeHandlePrimitive(T.tpe)
  .map { tree =>
    List((tree, Integer.valueOf(0)))
  }
  .getOrElse { ...
```